src/FirmwarePlugin/FirmwarePluginManager.cc

```diff
@@ -42,11 +42,46 @@ FirmwarePluginManager::~FirmwarePluginManager()
 
 }
 
-FirmwarePlugin* FirmwarePluginManager::firmwarePluginForAutopilot(MAV_AUTOPILOT autopilotType)
+FirmwarePlugin* FirmwarePluginManager::firmwarePluginForAutopilot(MAV_AUTOPILOT autopilotType, MAV_TYPE vehicleType)
 {
     switch (autopilotType) {
         case MAV_AUTOPILOT_ARDUPILOTMEGA:
-            return APMFirmwarePlugin::instance();
+            switch (vehicleType) {
+                case MAV_TYPE_QUADROTOR:
+                case MAV_TYPE_HEXAROTOR:
+                case MAV_TYPE_OCTOROTOR:
+                case MAV_TYPE_TRICOPTER:
+                    return ArduCopterFirmwarePlugin::instance();
+                    break;
+                // FIXME: The remainder of these need to be correctly assigned and new plugin classes created as needed.
```

Comment:
This needs to be completed by assigning the correct set to ArduCopter (I may have missed some). And also creating the correct sets for the other firmware types.